### PR TITLE
Add wireguard support

### LINKED ISSUE #1075: Does it support wireguard interfaces?
Hi,

I was just looking for a project like this :). 
Does it support wireguard interfaces? 
How can they be configured?

### PR DESCRIPTION
**Pull request by shreyagupta30**
This adds wireguard schema and it's tests. 

Signed-off-by : Shreya Gupta <shreyagupta3011@gmail.com>

### CODE REVIEW

**Comment by ovirt-infra** [comment]
Hello contributor, thanks for submitting a PR for this project! 

I am the bot who triggers "standard-CI" builds for this project.
As a security measure, I will not run automated tests on PRs that are not from white-listed contributors.

In order to allow automated tests to run, please ask one of the project maintainers to review the code and then do one of the following:

1. Type `ci test please` on this PR to trigger automated tests for it.
2. Type `ci add to whitelist` on this PR to trigger automated tests for it and also add you to the contributor white-list so that your future PRs will be tested automatically. ( keep in mind this list might be overwritten if the job XML is refreshed, for permanent whitelisting, please follow #3 option )
3. If you are planning to contribute to more than one project, maybe it's better to ask them to add you to the project organization, so you'll be able to run tests for all the organization's projects.

File: libnmstate/schema.py [hunk 1]
```diff
@@ -437,3 +439,14 @@ class Mode:
...
     CONFIG_SUBTREE = "mac-vtap"
+
+class Wireguard:
+    TYPE = InterfaceType.WIREGUARD
+    CONFIG_SUBTREE = "wireguard"
+    FWMARK = "fwmark" 
+    IPV4_AUTO_DEFAULT_ROUTE = "ipv4-auto-default-route"
+    IPV6_AUTO_DEFAULT_ROUTE = "ipv6-auto-default-route"	 
+    LISTEN_PORT = "listen-port" 	 	 
+    MTU = "mtu"
```
Comment (cathay4t) [inline libnmstate/schema.py]: The MTU is already included as `Interface.MTU`. There is no need to mention it here.

File: libnmstate/schema.py [hunk 2]
```diff
@@ -437,3 +439,14 @@ class Mode:
 class MacVtap(MacVlan):
     TYPE = InterfaceType.MAC_VTAP
     CONFIG_SUBTREE = "mac-vtap"
+
+class Wireguard:
+    TYPE = InterfaceType.WIREGUARD
+    CONFIG_SUBTREE = "wireguard"
+    FWMARK = "fwmark" 
+    IPV4_AUTO_DEFAULT_ROUTE = "ipv4-auto-default-route"
+    IPV6_AUTO_DEFAULT_ROUTE = "ipv6-auto-default-route"
```
Comment (cathay4t) [inline libnmstate/schema.py]: These two are just a wrapper done by NetworkManager.
The wireguard natively does not have them.

IMHO, user can manually get the same result using `fwmark` and `route rule` as mentioned in https://www.wireguard.com/netns/#routing-all-your-traffic

Please list the use cases which these options are required mandatory.

File: libnmstate/schema.py [hunk 3]
```diff
@@ -437,3 +439,14 @@ class Mode:
...
+
+class Wireguard:
+    TYPE = InterfaceType.WIREGUARD
+    CONFIG_SUBTREE = "wireguard"
+    FWMARK = "fwmark" 
+    IPV4_AUTO_DEFAULT_ROUTE = "ipv4-auto-default-route"
+    IPV6_AUTO_DEFAULT_ROUTE = "ipv6-auto-default-route"	 
+    LISTEN_PORT = "listen-port" 	 	 
+    MTU = "mtu" 	 	 
+    PEER_ROUTES = "peer-routes"
```
Comment (cathay4t) [inline libnmstate/schema.py]: The peer-routes can be done by existing route support of nmstate. No need to depend NM design for this.

**Review by cathay4t — CHANGES_REQUESTED** [review]
We are not child project of NetworkManager, please don't copy whatever NetworkManager provides to our schema. Nmstate has plugin design which allows other backends.

And nmstate schema should not depend on implementation of certain backend, so you should look on wireguard options itself instead of what NM provides.

The `ipv4-auto-default-route`, `peer-routes` and etc are just convenient stuffs created by NM which can be done via existing nmstate route and route rule schema.

You also missed a lot options:
 * Peer
 * KeepAlive
 * Allowed IP

We will not merge schema-only PR without implementation/test code showing it actual works.

We seems don't have intention to support wireguard server, please also mention that in commit message indication this is only for wireguard client.

**Comment by ffmancera** [comment]
> We are not child project of NetworkManager, please don't copy whatever NetworkManager provides to our schema. Nmstate has plugin design which allows other backends.
> 
> And nmstate schema should not depend on implementation of certain backend, so you should look on wireguard options itself instead of what NM provides.
> 

While I do agree on this, I see an issue here. NetworkManager provides certain options and some of them are wrappers that does not exist on wireguard. I will look at them and help @shreyagupta30 on this :-)

> The `ipv4-auto-default-route`, `peer-routes` and etc are just convenient stuffs created by NM which can be done via existing nmstate route and route rule schema.
> 
> You also missed a lot options:
> 
>     * Peer
> 
>     * KeepAlive
> 
>     * Allowed IP
> 

We are going to collect them and add these.

> 
> We will not merge schema-only PR without implementation/test code showing it actual works.

We are not planning on merging them now. This a PR for initial feedback which is nice.

> 
> We seems don't have intention to support wireguard server, please also mention that in commit message indication this is only for wireguard client.

Yep. That is missing.

By the way, if we want to support other options than the supported from NetworkManager we are going to need WIreguard support on Nispor and therefore, generic netlink support on rtnetlink. I will investigate this.

Thank you! @cathay4t @shreyagupta30

**Comment by ffmancera** [comment]
Not stale :-)

File: libnmstate/ifaces/ifaces.py [hunk 4]
```diff
@@ -44,6 +44,7 @@
 from .vlan import VlanIface
 from .vxlan import VxlanIface
 from .vrf import VrfIface
+from .wireguard import WireguardIface
```
Comment (thom311) [inline libnmstate/ifaces/ifaces.py:46]: the WireGuard maintainers have a strong opinion about the spelling and styling of WireGuard and the API.

In particular, they wish that `WireGuard` is spelled with upper `G`. NetworkManager's API does that too, at places where there is a CamelCase identifier (but it violates that are places like `NM_SETTING_WIREGUARD_SETTING_NAME` or `nm_wireguard_peer_new()`. By doing that, NetworkManager already violates it's own style, because from `NMSettingWireGuard` one would normally expect `NM_SETTING_WIRE_GUARD_SETTING_NAME`.

They probably would also like to see that the configuration options are spelled like why are in `man wg`. That is, `AllowedIPs` and `PresharedKey` instead of `allowed-ips` and `preshared-key`. However, that style already conflicted with NetworkManager's API, so NetworkManager does not do that either. So never mind that.

It's only about `WireGuard` instead of `Wireguard`.

Comment (shreyagupta30) [reply libnmstate/ifaces/ifaces.py:46]: Hi @thom311, I will make sure to correct it in my next commit. Thanks a lot for the review. :)

Comment (ffmancera) [reply libnmstate/ifaces/ifaces.py:46]: @cathay4t What do you think? Should we break our existing style? I do not like the idea.. but it's true that they have strong opinions on that.

**Comment by shreyagupta30** [comment]
Not stale :)

**Comment by ffmancera** [comment]
Not stale :.-)

**Comment by ffmancera** [comment]
Not stale again :(

**Comment by shreyagupta30** [comment]
Not stale

File: libnmstate/ifaces/ifaces.py (hunk 4 above)
Comment (cathay4t) [reply libnmstate/ifaces/ifaces.py:46]: For internal stuff like class name, you can follow the guidance of WireGuard. But for public property, please use lower case with dash for consistency

File: libnmstate/nm/ieee_802_1x.py [hunk 5]
```diff
@@ -17,11 +17,11 @@
 # along with this program. If not, see <https://www.gnu.org/licenses/>.
 #
 
-from libnmstate.error import NmstateLibnmError
 from libnmstate.schema import Ieee8021X
 
 from .common import NM
 from .common import GLib
+from .utils import get_secrets
```
Comment (cathay4t) [inline libnmstate/nm/ieee_802_1x.py]: If we have file called `utils.py`, it will eventually grow bloated. Please use `secret.py` instead.

File: libnmstate/nm/wireguard.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        for allowed_ip in peer.get(Wireguard.Peer.ALLOWED_IPS, []):
+            wireguard_peer.append_allowed_ip(allowed_ip, True)
+
+        wireguard_setting.append_peer(wireguard_peer)
+
+
+def get_info(context, nm_ac):
+    if not nm_ac:
+        return {}
+    nm_profile = nm_ac.get_connection()
```
Comment (cathay4t) [inline libnmstate/nm/wireguard.py:78]: Even this is for temporary use(we will move to Nispor for querying the kernel interface), but please use `applied_config` if possible. Please check `nm/ipv4.py` for example.

File: libnmstate/nm/connection.py [hunk 7]
```diff
@@ -142,7 +143,6 @@ def create_new_nm_simple_conn(iface, nm_profile):
 
         if wired_setting:
             settings.append(wired_setting)
-
```
Comment (cathay4t) [inline libnmstate/nm/connection.py]: This is unrelated change.

File: libnmstate/nm/utils.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    return secrets
+
+
+def _get_secrets_callback(nm_profile, result, user_data):
+    context, secrets, action, setting_name = user_data
+
+    try:
+        nm_secrets = nm_profile.get_secrets_finish(result)
+    except GLib.Error as e:
+        context.fail(NmstateLibnmError(f"{action} failed: error={e}"))
```
Comment (cathay4t) [inline libnmstate/nm/utils.py]: A `return` is required here.

File: libnmstate/schema.py [hunk 9]
```diff
@@ -459,3 +461,19 @@ class Ieee8021X:
...
     CA_CERT = "ca-cert"
+
+
+class Wireguard:
+    TYPE = InterfaceType.WIREGUARD
+    CONFIG_SUBTREE = "wireguard"
+    FWMARK = "fwmark"
+    LISTEN_PORT = "listen-port"
+    PRIVATE_KEY = "private-key"
+    PEERS = "peers"
```
Comment (cathay4t) [inline libnmstate/schema.py]: The `Wireguard.PEERS` should changed to `Wireguard.Peer.CONFIG_SUBTREE` for consistency.

File: libnmstate/schemas/operational-state.yaml [hunk 10]
```diff
@@ -716,6 +717,45 @@ definitions:
...
+                  endpoint:
+                    type: string
+                  persistent-keepalive:
+                    type: integer
+                  preshared-key:
+                    type: string
+                  public-key:
+                    type: string
+                  flags:
+                    type: integer
```
Comment (cathay4t) [inline libnmstate/schemas/operational-state.yaml]: Both `flags` and `peer.flags` are not used. Please remove them.

File: libnmstate/schemas/operational-state.yaml [hunk 11]
```diff
@@ -716,6 +717,45 @@ definitions:
...
+      properties:
+        type:
+          type: string
+          enum:
+            - wireguard
+        wireguard:
+          type: object
+          properties:
+            fwmark:
+              type: integer
```
Comment (cathay4t) [inline libnmstate/schemas/operational-state.yaml]: The wireguard wg CLI tool supports string like `0xf1`, please support it also.

File: tests/integration/wireguard_test.py [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+
+
+def test_modify_wireguard():
+    with wireguard_interface(WG0) as desired_state:
+        wg_state = desired_state[Interface.KEY][0][Wireguard.CONFIG_SUBTREE]
+        wg_state[Wireguard.PRIVATE_KEY] = WG1_PRIVATE_KEY
+        wg_state[Wireguard.LISTEN_PORT] = 9500
+
+        libnmstate.apply(desired_state)
+        assertlib.assert_state(desired_state)
```
Comment (cathay4t) [inline tests/integration/wireguard_test.py:57]: Please use `assert_state_match`, in the future there might be more properties added to wireguard, the `assert_state` will fail then.

File: tests/integration/wireguard_test.py [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+        wg_state[Wireguard.PEERS] = [
+            {
+                Wireguard.Peer.ALLOWED_IPS: ["192.0.2.1"],
+                Wireguard.Peer.ENDPOINT: "my-wg.example.com:4001",
+                Wireguard.Peer.PRESHARED_KEY: PEER0_PRESHARED_KEY,
+                Wireguard.Peer.PUBLIC_KEY: PEER0_PUBLIC_KEY,
+            },
+        ]
+        libnmstate.apply(desired_state)
+        assertlib.assert_state_match(desired_state)
```
Comment (cathay4t) [inline tests/integration/wireguard_test.py:87]: Please use `assert_state_match` instead.

File: tests/integration/wireguard_test.py [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+
+
+def test_remove_wireguard_peer():
```
Comment (cathay4t) [inline tests/integration/wireguard_test.py:94]: Currently, you are trusting `libnmstate.apply()` to do the proper verification. But your patch is ask NM to write a profile, then we verify it by reading this NM profile in verification stage. That does not sounds like trustable.

Could you consider to use `wg showconf` to actual check whether we are doing correct?

File: libnmstate/ifaces/wireguard.py [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+#
+
+from libnmstate.schema import Wireguard
+
+from .base_iface import BaseIface
+
+
+class WireguardIface(BaseIface):
```
Comment (cathay4t) [inline libnmstate/ifaces/wireguard.py:25]: Need unit test case for this class.

**Review by cathay4t — CHANGES_REQUESTED** [review]
In summery, these changes required:
 * Rename `utils.py` to `secret.py` to be more specific.
 * Support hex string for `fwmark`.
 * Improve the test to use `wg showconf` for validation.
 * Need unit test case for `WireguardIface`.

**Comment by shreyagupta30** [comment]
not stale

File: libnmstate/nm/wireguard.py (hunk 6 above)
Comment (ffmancera) [reply libnmstate/nm/wireguard.py:78]: Hi Gris, this won't work. applied_config is a `NM.SimpleConnection` and that class does not contain `get_secrets_async()` that is needed to get the secrets of Wireguard. The class required is `NM.RemoteConnection` which is the one that `nm_ac.get_connection()` is provided. I've been working with Shreya on this but I think unfortunately we cannot use `applied_config` here.

**Comment by shreyagupta30** [comment]
Not stale

**Comment by ffmancera** [comment]
This was not stale, we are looking at it. Sorry for the inconvenience.

File: libnmstate/nm/wireguard.py (hunk 6 above)
Comment (ffmancera) [reply libnmstate/nm/wireguard.py:78]: What do you think about my comment @cathay4t?

Comment (cathay4t) [reply libnmstate/nm/wireguard.py:78]: If applied connection does not provide such info, I am OK with saved configure.
